dashboard/app/models/user_level.rb

```diff
@@ -67,42 +67,71 @@ def passing?
     ActivityConstants.passing?(best_result)
   end
 
-  # user levels can be linked through pair programming. The 'driver'
-  # user level is the one that is linked to the user account that
-  # completed the activity; the 'navigator' user level is the one that
-  # also gets credit for the solution.
-  has_many :paired_user_levels_as_navigator, class_name: 'PairedUserLevel', foreign_key: 'navigator_user_level_id'
-  has_many :paired_user_levels_as_driver, class_name: 'PairedUserLevel', foreign_key: 'driver_user_level_id'
-
-  has_many :navigator_user_levels, through: :paired_user_levels_as_driver
-  has_many :driver_user_levels, through: :paired_user_levels_as_navigator
+  # Retrieves and memoizes the latest PairedUserLevel that's associated with
+  # this UserLevel for internal callers. External callers should call one of
+  # the higher-level pairing-related methods below.
+  #
+  # Conceptually, each UserLevel should only be associated with one
+  # PairedUserLevel. However, since we don't clean up previous entries in the
+  # paired_user_levels table when posting progress, there can be multiple
+  # entries in the paired_user_levels table associated with a user_level. We
+  # heuristically consider the latest entry in the paired_user_levels table as
+  # the "active" one. This is correct for most cases but can be incorrect in
+  # some edge cases such as when a student leaves a pairing group and makes
+  # further progress on a level as an individual.
+  private def latest_paired_user_level
+    return @latest_paired_user_level if defined? @latest_paired_user_level
+    @latest_paired_user_level =
+      PairedUserLevel.where(driver_user_level_id: id).
```

Comment:
Does `PairedUserLevel.where(driver_user_level_id: id).or(PairedUserLevel.where(navigator_user_level_id: id))` return them ordered by latest entry last?

Reply:
This query returns values in the default order which is by id.  This is effectively the same as ordering by creation/modification time (ascending) because entries in this table are never modified.